hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3136,6 +3139,26 @@ def save(self, uri):
 
         Env.backend().execute(NDArrayWrite(self._ir, hl.str(uri)._ir))
 
+    def to_numpy(self):
+        """Execute and convert this NDArray to a `NumPy ndarray.
+
+        Examples
+        --------
+        >>> a = nd.to_numpy() # doctest: +SKIP
+
+        Notes
+        -----
+        The resulting ndarray will have the same shape as the block matrix.
+
+        Returns
+        -------
+        :class:`numpy.ndarray`
+        """
+        # FIXME Use filesystem abstraction instead when that is ready
+        temp_file = tempfile.NamedTemporaryFile(suffix='.npy').name
+        self.save(temp_file)
+        return np.load(temp_file)
```

Comment:
I need to spend a bit of time learning how all this stuff works. 

Looking at `save` above, I see this is a method on generic NDArray expressions, but it will only work (a) on NDArrays whose IR does not contain a reference to a field of a MatrixTable or Table (will crash on a global field, and totally undefined on a Table/MatrixTable distributed field), and (b) on NDArrays of type float64 (I believe. I haven't investigated the .npy format but numpy doesn't know how to decode arbitrary Hail types).

At the very least, we should add these checks, because users will get weird errors if we don't. We should also move `save` onto NDArrayNumericExpression with an implicit conversion to float64 before executing, right?

However, I don't think we should be putting methods on any kind of `Expression` that only work on a certainly category of expressions -- a method should work on all instances of the class.

Reply:
Ya you're right I'm not sure how this will behave in distributed contexts. It's not necessarily just float64s though. NDArrayNumericExpression could be more reasonable because numerics are the only types I've implemented so far, but the format is not strictly numeric.

Reply:
But yes we can't just slam hail values in there

Reply:
I can't find anything specifically about numpy compatibility. It just relies on the fact that the 

```
  val unblockedUncompressed = new PackCodecSpec(new StreamBufferSpec)
```

writes out NDArrays using the fundamental struct/array representation that just happens to be numpy compatible, right?

Reply:
`EncodeNDArray.scala` specifies which element types we can encode in the `.npy` format. Not sure exactly what you mean by "fundamental struct/array" but our representation for structured types is not immediately numpy compatible

Reply:
ahh, there's a separate cxx pack encoder. that explains it.

Reply:
okay, yeah, we just need to move `save` onto the ndarray numeric expression.

Reply:
Sounds good!